File: src/Dstar.cpp
#include <stdio.h>
#include "Dstar.hpp"
#include <algorithm>
#include <cmath>

#include <ros/ros.h>

std::vector<coordinate> Dstar::init_path() {
    cell * goal_c = this->get_ptr(goal);
    *goal_c = cell(0, 0, goal_c, Tag::Open, goal);
    
    while (1) {
        get_kmin_result k_min = this->process_state();
        cell * start_c = this->get_ptr(start);
        if ((k_min.exists == false) || (start_c->t == Tag::Closed)) {
            return this->get_path(start_c);
        }
    }
}

void Dstar::change_map(octomap::OcTree * tree) {
    //auto center = octomap::point3d(0, 0, 0);
    //auto half_size = tree->getNodeSize(0) / 2;
    //auto min_bbx = octomap::point3d(center.x() - half_size, center.y() - half_size, height - 0.5);
    //auto max_bbx = octomap::point3d(center.x() + half_size, center.y() + half_size, height + 0.5);
    for (auto leaf_bbx_it = tree->begin_leafs(16);
            leaf_bbx_it != tree->end_leafs();
            leaf_bbx_it++) {
        // Find current bounds, iterate through affected cells
        auto leaf_center = leaf_bbx_it.getCoordinate();
        auto leaf_half_size = leaf_bbx_it.getSize() / 2;

        int min_x = tree->coordToKey(leaf_center.x() - leaf_half_size);
        int max_x = tree->coordToKey(leaf_center.x() + leaf_half_size);
        int min_y = tree->coordToKey(leaf_center.y() - leaf_half_size);
        int max_y = tree->coordToKey(leaf_center.y() + leaf_half_size);

        min_x = std::max(min_x - offset, 0);
        max_x = std::min(max_x - offset, size.x);
        min_y = std::max(min_y - offset, 0);
        max_y = std::min(max_y - offset, size.y);

        for (int j = min_y; j < max_y; j++) {
            for (int i = min_x; i < max_y; i++) {
                if (!this->occupancy[i + j * size.x] < leaf_bbx_it->getOccupancy()) {
                    // Update new cost
                    this->add_obstacle(coordinate(i, j), leaf_bbx_it->getOccupancy());
                }
            }
        }
    }
}

void Dstar::add_obstacle(coordinate loc, double new_probability) {
    ROS_INFO("Updated obstacle at (%d, %d): %f", loc.x, loc.y, new_probability * 1000);

    double diff = new_probability - this->occupancy[loc.x + loc.y * size.x];

    int min_x = std::max(0, loc.x - 5);
    int max_x = std::min(size.x, loc.x + 6);
    int min_y = std::max(0, loc.y - 5);
    int max_y = std::min(size.y, loc.y + 6);

    double max_distance = std::sqrt(5.0 * 5 * 2);

    for (int j = min_y; j < max_y; j++) {
        for (int i = min_x; i < max_x; i++) {
            double distance = std::sqrt(i * i + j * j);
            double multiplier = 1000.0 - distance * 1000 / max_distance;
            this->costs[i + j * size.x] += diff * multiplier;

            cell * curr = this->get_ptr(coordinate(i, j));
            if (curr->t == Tag::Closed) {
                this->insert(curr, curr->h);
            }
        }
    }
}

std::vector<coordinate> Dstar::navigate_map(coordinate curr) {
    cell * curr_cell_ptr = this->get_ptr(curr);
    while(1) {
        get_kmin_result k_min = this->process_state();
        
        if (((k_min.exists == false) ||
                    (curr_cell_ptr->h <= k_min.kmin)) &&
                !(curr_cell_ptr->t == Tag::Open)) {
            return this->get_path(curr_cell_ptr);
        }
    }
}

void Dstar::insert(cell * curr, double new_h) {
    if (curr->t == Tag::New) {
        curr->h = new_h;
        curr->k = new_h;
        curr->t = Tag::Open;
        this->put_open(curr);
    } else if (curr->t = Tag::Open) {
        curr->h = new_h;
        curr->k = std::min(curr->k, new_h);

        std::make_heap(pqueue.begin(), pqueue.end());
    } else if (curr->t == Tag::Closed) {
        curr->h = new_h;
        curr->k = std::min(curr->h, new_h);
        curr->t = Tag::Open;
        this->put_open(curr);
    }
}

get_kmin_result Dstar::process_state() {
    auto curr = this->get_open();
    
    if (curr == NULL) {
        get_kmin_result result;
        result.exists = false;
        result.kmin = -1.0;
        return result;
    }

    if (curr->k < curr->h) {
        auto neighbors = this->get_neighbors(curr->loc);
        for (auto neighbor_it = neighbors.begin(); neighbor_it < neighbors.end(); neighbor_it++) {
            auto neighbor = *neighbor_it;
            if (neighbor->t != 'n' &&
                    neighbor->h <= curr->k &&
                    curr->h > neighbor->h + this->get_cost(neighbor->loc, curr->loc)) {
                curr->b = neighbor;
                curr->h = neighbor->h + this->get_cost(neighbor->loc, curr->loc);
            }
        }
    }

    if (curr->k == curr->h) {
        auto neighbors = this->get_neighbors(curr->loc);
        for (auto neighbor_it = neighbors.begin(); neighbor_it < neighbors.end(); neighbor_it++) {
            auto neighbor = *neighbor_it;
            if ((neighbor->t == 'n') ||
                    (neighbor->b == curr && neighbor->h != curr->h + this->get_cost(curr->loc, neighbor->loc)) ||
                    (neighbor->b != curr && neighbor->h > curr->h + this->get_cost(curr->loc, neighbor->loc))) {
                neighbor->b = curr;
                this->insert(neighbor, curr->h + this->get_cost(curr->loc, neighbor->loc));
            }
        }
    } else {
        auto neighbors = this->get_neighbors(curr->loc);
        for (auto neighbor_it = neighbors.begin(); neighbor_it < neighbors.end(); neighbor_it++) {
            auto neighbor = *neighbor_it;
            if (neighbor->t == 'n' ||
                    (neighbor->b == curr && neighbor->h != curr->h + this->get_cost(curr->loc, neighbor->loc))) {
                neighbor->b = curr;
                this->insert(neighbor, curr->h + this->get_cost(curr->loc, neighbor->loc));

            } else if (neighbor->b != curr && neighbor->h > curr->h + this->get_cost(curr->loc, neighbor->loc)) {
                this->insert(curr, curr->h);
            } else if (neighbor->b != curr && curr->h > neighbor->h + this->get_cost(curr->loc, neighbor->loc) &&
                    neighbor->t == 'c' && neighbor->h > curr->k) {
                this->insert(neighbor, neighbor->h);
            }
        }
    }

    return this->get_kmin();
}

Dstar::Dstar(coordinate start, coordinate goal, double height) :
        start(start), goal(goal), height(height) {
    size = coordinate(65536/16, 65536/16);
    world = new cell[size.x * size.y];
    costs = new double[size.x * size.y];
    std::fill(costs, costs + size.x * size.y, 0.0);

    occupancy = new double[size.x * size.y];
    std::fill(occupancy, occupancy + size.x * size.y, 0.0);

    offset = (65536 - size.x) / 2;
}

Dstar::~Dstar() {
    delete world;
    delete costs;
    delete occupancy;
}

cell Dstar::get(coordinate n) {
    return world[n.x + n.y * size.x];
}

cell * Dstar::get_ptr(coordinate n) {
    return &world[n.x + n.y * size.x];
}

void Dstar::put(coordinate n, cell curr) {
    *this->get_ptr(n) = curr;   
}

cell * Dstar::get_open() {
    if (!pqueue.empty()) {
        auto curr = pqueue.front();
        std::pop_heap(pqueue.begin(), pqueue.end());
        pqueue.pop_back();
        curr->t = Tag::Closed;
        return curr;
    } else {
        return NULL;
    }
}

void Dstar::put_open(cell * curr) {
    curr->t = Tag::Open;
    // open_cells
    
    pqueue.push_back(curr);
    std::push_heap(pqueue.begin(), pqueue.end());
}

get_kmin_result Dstar::get_kmin() {
    get_kmin_result result;
    if (!pqueue.empty()) {
        auto curr = pqueue.front();
        result.exists = true;
        result.kmin = curr->k;
        return result;
    }
    else {
        result.exists = false;
        result.kmin = -1;
        return result;
    }
}

double Dstar::get_cost(coordinate loc1, coordinate loc2) {
    // Unsure what structure of cost
    int diff_x = loc1.x - loc2.x;
    int diff_y = loc1.y - loc2.y;
    double distance = std::sqrt((double)(diff_x * diff_x + diff_y * diff_y));
    return costs[loc2.x + loc2.y * size.x] + distance;
}

std::vector<coordinate> Dstar::get_path(cell *curr) {
    std::vector<coordinate> path;
    path.push_back(curr->loc);
    
    cell * temp = curr;
    while (1) {
        temp = temp->b;
        
        if (temp == NULL) {
            return std::vector<coordinate>(0);
        }
        else {
            path.push_back(temp->loc);
            
            if ((temp->loc.x == goal.x) && (temp->loc.y == goal.y)) {
                return path;
            }
        }          
    }
}

std::vector<cell *> Dstar::get_neighbors(coordinate loc) {
    std::vector<cell *> neighbors;
    int x = loc.x;
    int y = loc.y;
    
    int min_y = std::max(0, y - 1);
    int max_y = std::min(size.y - 1, y + 1);
    int min_x = std::max(0, x - 1);
    int max_x = std::min(size.x - 1, x + 1);
    
    for (int j = min_y; j < max_y + 1; j++) {
        for (int i = min_x; i < max_x + 1; i++) {
            if (i == x && j == y) { continue; }
            coordinate loc;
            loc.x = i;
            loc.y = j;
            neighbors.push_back(get_ptr(loc));
        }
    }
    return neighbors;
}

File: src/Dstar.py
# -*- coding: utf-8 -*-
"""
Created on Thu Mar 16 11:04:31 2017

@author: Sith, Andrew, Evan
"""

import sys
import math
import Queue
from graphics import Graphics

class Cell:

    # Access cell values:
    #   h:      path cost
    #   k:      smallest value of h seen so far
    #   b:      back pointer
    #   t:      tag ('c' - closed; 'o' - open; 'n' - new)
    def __init__(self, h, k, b, t, loc):
        self.h = h
        self.k = k
        self.b = b
        self.t = t
        self.loc = loc


class D_Star:
    
    # Init function with robot world's decomposition
    # Start (x,y) and Goal (x,y) 
    def __init__(self, size, start, goal):
        world_x = size[0]
        world_y = size[1]
        
        self.start = start
        self.goal = goal
        self.size = size
        self.world = [[Cell(None, None, None, 'n', (i, j)) for i in range(world_x)] for j in range(world_y)]
        self.pqueue = Queue.PriorityQueue() # Open list
        self.costs = [[[[None for _ in range(world_x)] for _ in range(world_y)] for _ in range(world_x)] for _ in range(world_y)]
        self.obstacles = []

        for j1 in range(world_y):
            for i1 in range(world_x):
                for j2 in range(world_y):
                    for i2 in range(world_x):
                        self.costs[j1][i1][j2][i2] = math.hypot(i2 - i1, j2 - j1)

    # Helper functions below
    # n denotes some tuple (i, j)
    # curr denotes some Cell object

    def get(self, n):
        return self.world[n[1]][n[0]]

    def put(self, n, cell):
        self.world[n[1]][n[0]] = cell

    def get_open(self):
        if not self.pqueue.empty():
            item = self.pqueue.get()
            curr = self.get(item[1])
            curr.t = 'c'

            return (item[0], curr)

        else:
            return (None, None)

    def put_open(self, cell):
        cell.t = 'o'
        
        self.pqueue.put((cell.h, cell.loc))

    def get_kmin(self):
        try:
            return sorted(self.pqueue.queue)[0][0]
        except:
            print "Unexpected error:", sys.exc_info()[0]
            return None

    def get_cost(self, curr1, curr2):
        x1, y1 = curr1.loc
        x2, y2 = curr2.loc

        return self.costs[y2][x2][y1][x1]

    def get_path(self, curr):
        path = [curr.loc]

        print "getting path from {}".format(curr.loc)
        print "open list: {}".format(self.pqueue.queue)

        while True:
            curr = curr.b

            if curr == None:
                return None
            else:
                print curr.loc
                if curr.loc in path:
                    raise Exception('Recursive backlinks')
                path.append(curr.loc)

                if curr.loc == self.goal:
                    return path

    def get_neighbors(self, curr):
        neighbors = []
        loc = curr.loc
        x = loc[0]
        y = loc[1]

        min_y = max(0, y - 1)
        max_y = min(self.size[1] - 1, y + 1)
        min_x = max(0, x - 1)
        max_x = min(self.size[0] - 1, x + 1)                
        
        for j in xrange(min_y, max_y + 1):
            for i in xrange(min_x, max_x + 1):
                if (i, j) == (x, y):
                    continue
                neighbors.append(self.get((i, j)))

        return neighbors

    ##################################################

    def init_path(self):
        goal = Cell(0, 0, 'g', 'o', self.goal)
        goal.b = goal
        self.put(self.goal, goal)
        self.put_open(goal)

        while True:
            k_min = self.process_state()
            print "Got k_min: {}".format(k_min)
            start = self.get(self.start)

            if k_min == None or start.t == 'c':
                return self.get_path(start)

    def change_map(self, actual_map, curr):
        for sensed in self.get_neighbors(curr):
            i, j = sensed.loc
            if actual_map[j][i]:
                self.insert(sensed, sensed.k)
                for neighbor in self.get_neighbors(sensed):
                    if self.costs[j][i][neighbor.loc[1]][neighbor.loc[0]] != float("inf"):
                        if not (i, j) in self.obstacles:
                            print "found obstacle: {}".format((i, j))
                            self.obstacles.append((i, j))
                        self.modify_costs(neighbor, sensed, float("inf"))

    def navigate_map(self, curr):
        while True:
            k_min = self.process_state()

            print "Got k_min: {}".format(k_min)

            if (k_min == None or curr.h <= k_min) and not curr.t == 'o':
                return self.get_path(curr)

    def modify_costs(self, curr1, curr2, new_c):
        x1, y1 = curr1.loc
        x2, y2 = curr2.loc
        self.costs[y2][x2][y1][x1] = new_c
        self.costs[y1][x1][y2][x2] = new_c

        if curr1.t == 'c':
            self.insert(curr1, curr1.h)

        return self.get_kmin()

    def insert(self, curr, h_new):
        print "inserting {} {} {}".format(curr.loc, curr.t, h_new)
        if curr.t == 'n':
            curr.h = h_new
            curr.k = h_new
            curr.t = 'o'
            self.put_open(curr)
        elif curr.t == 'o':
            curr.h = h_new
            curr.k = min(curr.k, h_new)
            indexes = [i for i,x in enumerate(self.pqueue.queue) if x[1] == curr.loc]
            for index in indexes:
                self.pqueue.queue[index] = (curr.k, curr.loc)
        elif curr.t == 'c':
            curr.h = h_new
            curr.k = min(curr.h, h_new)
            curr.t = 'o'
            self.put_open(curr)
    

    def process_state(self):
        k_old, curr = self.get_open()
        
        if k_old == None:
            print "popped {} from the open queue".format(k_old)
            return None

        print "popped ({}, {}) from the open queue with h: {}".format(k_old, curr.loc, curr.h)

        if k_old < curr.h:
            print "Raise {}".format(curr.loc)
            for neighbor in self.get_neighbors(curr):
                if neighbor.t != 'n' and neighbor.h <= k_old \
                and curr.h > neighbor.h + self.get_cost(neighbor, curr):
                    curr.b = neighbor
                    curr.h = neighbor.h + self.get_cost(neighbor, curr)

        if k_old == curr.h:
            for neighbor in self.get_neighbors(curr):
                if (neighbor.t == 'n') \
                or (neighbor.b == curr and neighbor.h != curr.h + self.get_cost(curr, neighbor)) \
                or (neighbor.b != curr and neighbor.h > curr.h + self.get_cost(curr, neighbor)):
                    neighbor.b = curr
                    self.insert(neighbor, curr.h + self.get_cost(curr, neighbor))

        else:
            for neighbor in self.get_neighbors(curr):
                if neighbor.t == 'n' \
                or (neighbor.b == curr and neighbor.h != curr.h + self.get_cost(curr, neighbor)):
                    neighbor.b = curr
                    self.insert(neighbor, curr.h + self.get_cost(curr, neighbor))

                elif neighbor.b != curr and neighbor.h > curr.h + self.get_cost(curr, neighbor):
                    self.insert(curr, curr.h)

                elif neighbor.b != curr and curr.h > neighbor.h + self.get_cost(curr, neighbor) \
                and neighbor.t == 'c' and neighbor.h > k_old:
                    self.insert(neighbor, neighbor.h)

        return self.get_kmin()

    def run(self, actual_map, size):
        final_path = [self.start]
        graphics = Graphics(actual_map, self.start, self.goal, size=size)
        graphics.display(final_path, self.obstacles)

        path = self.init_path()
        if path == None:
            return None

        curr = self.get(self.start)

        while curr.loc != self.goal:
            self.change_map(actual_map, curr)
            path = self.navigate_map(curr)

            if path == None:
                return None

            curr = self.get(path[1])
            final_path.append(path[1])

            graphics.display(final_path, self.obstacles)

        graphics.display(final_path, self.obstacles, done=True)
        return final_path
